### ISSUE TITLE
Bally / Midway MCR games

### ISSUE THREAD

**Issue by arcadez** [arcadez/mame2003-plus-libretro#34]
https://github.com/libretro/mame2003-libretro/issues/121

**Comment by arcadez** [comment]
It might be worth making a build from this point in the project and testing a the MCR games

https://github.com/arcadez/mame2003-plus-libretro/tree/7da127055ce603d5c1bba81127a994ae6c8ee9e7

**Comment by markwkidd** [comment]
Do you have some kind of device that can run RetroArch other than your xbox. No kind of desktop, laptop, or tablet? 

I took a minute to try to build an old version and I'm

**Comment by arcadez** [comment]
Ah i forgot that you can make a RetroArch PC build, i'll need to look into what tools would be
needed sometime as i'd like to run a few debugs with the current core to see what is happening
exactly with those Midway MCR games eg why some of them lock into service mode on boot 
while some others actually start but refuse to coin.

Plus im also keen to see if they actually would work when using an older source

**Comment by markwkidd** [comment]
It's very easy to use RetroArch on PC. Glad to help if I can. Download RetroArch here: http://www.retroarch.com/?page=platforms

At this point you can start RetroArch and install the most recent mame2003-plus binary via the "Online Updater" which downloads cores that the buildbot has compiled.

If you want to compile directly for the PC version of RetroArch I should be able to help you get going there if you like. There are also pretty good docs on compilation here https://docs.libretro.com/ although I don't know if you will need all that information since you have experience already

**Comment by arcadez** [comment]
Thanks for the links i will get around to this at some point but im busy with another project
at the moment backporting the latest Taito C-Chip MCU code for a number of games.....

[img]https://mamedev.emulab.it/haze/2018/03/20/now-you-c-inside/[/img]

**Comment by markwkidd** [comment]
ah hah! no rush, but let me know if I can be your RetroArch helper -- it's the least I can do after all your backport work

**Comment by grant2258** [comment]
whats your hunch on this one arcadez?

**Comment by arcadez** [comment]
Hmm i seem to have it in my head either rightly or wrongly that some corewide input changes
are to blame here as the games are locking into service mode on boot or refuse to coin this
to me points to an input problem as dips and inputs are closely aligned.

These games tend to have their inputs and coins handled by the main CPU in this case a Z80
via the pia6821 machine emulated device the problem could lie here right enough.
im gonna take the latest source and rollback the input changes to before july 2016 
and see if i can then make a test build.

**Comment by grant2258** [comment]
I'll do after work arcadez compiling on i7 with add on takes a few seconds

**Comment by grant2258** [comment]
ok got it compiled arcades mc3 or mcr2 what games you need tested?

**Comment by grant2258** [comment]
well changes i can tell you is rampage now goes to a black screen but you can use the tab menu.  tapper goes to that pallete screen that  rampage goes to after  the inpt.c and h from the other thread. I think it is related to this code doesnt mean a 100~% it is though will need looked into more i imagine

**Comment by grant2258** [comment]
one note of intrest a/grant/empty/Download/Emulation/roms/arcade/MAME2003 None Merged//mame2003/nvram/tapper.nv])

notice the double // for the nvram that could be causing the problems as well

**Comment by grant2258** [comment]
that in itself is probably easy enough to fix somthing else is calling osd_get_path_info with  ".zip" as filename will need to check out if this is normal  behaviour or not

**Comment by grant2258** [comment]
well i for back to basics i reverted too git reset --hard 82ef12f78c this is 2012 and the input system was as it should have been then and tapper accepts credits.  The inptport.c and .h should be messed with as little as possible

you might have trouble getting roms to load just do this

strace -o logfile -e trace=file /home/grant/ra/usr/local/bin/retroarch

try start tapper

close reatroach 

less logfile | grep tapper.zip

copy the rom where it should be and enjoy it working

**Comment by grant2258** [comment]
oh yes rampage works as well :). I not touching this one at all I know there is going to be big discussions on default setups ect. 

I personally think we should just define the retropad buttons in mame something like button 1 = retropad a button 2= b in some config file and do an onload config change for sf2. Just my 2 pennies worth

**Comment by grant2258** [comment]
this is where we broke midway input system https://github.com/libretro/mame2003-libretro/commit/497e5173bfbd10c1514c0e8c669c261adf02e784

**Comment by arcadez** [comment]
Nice work tracking down the cause!! alotta folks will be happy to hear about this.

**Comment by dankcushions** [comment]
strange, that commit doesn't seem to have anything to do with input, but rather audio? unless it's causing some timing issue

**Comment by grant2258** [comment]
well the input comes from the audio cpu  maybe the timing is off and we are loosing chunks of data or its just out of sync

**Comment by grant2258** [comment]
I looked at the code more didnt want to compile and say yes I followed what the osd_audio stuff was doing. I think that machine driver is internal to the game and we shouldnt be using it could explain the timing problems see src/cpuexec.c

**Comment by grant2258** [comment]
/*************************************
 *
 *      Machine drivers
 *
 *************************************/

/* Core MCR3 system with no sound */
static MACHINE_DRIVER_START( mcr3 )

        /* basic machine hardware */
        MDRV_CPU_ADD_TAG("main", Z80, 5000000)
        MDRV_CPU_CONFIG(mcr_daisy_chain)
        MDRV_CPU_MEMORY(readmem,writemem)
        MDRV_CPU_PORTS(readport,writeport)
        MDRV_CPU_VBLANK_INT(mcr_interrupt,2)

        MDRV_FRAMES_PER_SECOND(30)
        MDRV_VBLANK_DURATION(DEFAULT_REAL_30HZ_VBLANK_DURATION)
        MDRV_MACHINE_INIT(mcr)
        MDRV_NVRAM_HANDLER(mcr3)

        /* video hardware */
        MDRV_VIDEO_ATTRIBUTES(VIDEO_TYPE_RASTER | VIDEO_UPDATE_BEFORE_VBLANK)
        MDRV_SCREEN_SIZE(32*16, 30*16)
        MDRV_VISIBLE_AREA(0*16, 32*16-1, 0*16, 30*16-1)
        MDRV_GFXDECODE(gfxdecodeinfo)
        MDRV_PALETTE_LENGTH(64)

        MDRV_VIDEO_START(mcr3)
        MDRV_VIDEO_UPDATE(mcr3)
MACHINE_DRIVER_END

**Comment by grant2258** [comment]
think this could be linked to the change that was made that broke mcr3

https://github.com/libretro/mame2003-libretro/issues/75

in line 911  in mame.c  I dont think we should be doing this at all the buffer should be handling its own under and over runs to keep it close to what it should be might be ok though im pretty sure the emulator changes these values to sync up when it all goes wrong

**Comment by dankcushions** [comment]
> think this could be linked to the change that was made that broke mcr3
> libretro/mame2003-libretro#75

can you clarify how?

**Comment by grant2258** [comment]
I think the audio sync problem is the way we are handling the in the osd_audio_update the mame core. I dont know much about the retropie api at all though. Im just pushing some thoughts the only place i really see needing set to the game refresh rate is already done at 

./src/mame.c:717:       params.fps = Machine->drv->frames_per_second;

**Comment by grant2258** [comment]
Im going to catch some zzs before work will try looking into this propely when i get some days off. Ill need to compile the core at the point it works again and test them games and see if the video is good

**Comment by grant2258** [comment]
//============================================================
//	osd_update_audio_stream
//============================================================

int osd_update_audio_stream(INT16 *buffer)
{
	// if nothing to do, don't do it
	if (Machine->sample_rate != 0 && stream_buffer)
	{
		int original_bytes = bytes_in_stream_buffer();
		int input_bytes = samples_this_frame * stream_format.nBlockAlign;
		int final_bytes;

		// update the sample adjustment
		update_sample_adjustment(original_bytes);

		// copy data into the sound buffer
		copy_sample_data(buffer, input_bytes);

		// check for overflows
		final_bytes = bytes_in_stream_buffer();
		if (final_bytes < original_bytes)
			buffer_overflows++;
	}

	// reset underflow/overflow tracking
	if (++total_frames == INGORE_UNDERFLOW_FRAMES)
		buffer_overflows = buffer_underflows = 0;

	// update underflow/overflow logging
#if (DISPLAY_UNDEROVERFLOW || LOG_SOUND)
{
	static int prev_overflows, prev_underflows;
	if (total_frames > INGORE_UNDERFLOW_FRAMES && (buffer_overflows != prev_overflows || buffer_underflows != prev_underflows))
	{
		prev_overflows = buffer_overflows;
		prev_underflows = buffer_underflows;
#if DISPLAY_UNDEROVERFLOW
		usrintf_showmessage("overflows=%d underflows=%d", buffer_overflows, buffer_underflows);
#endif
#if LOG_SOUND
		fprintf(sound_log, "************************ overflows=%d underflows=%d\n", buffer_overflows, buffer_underflows);
#endif
	}
}
#endif

	// compute how many samples to generate next frame
	samples_left_over += samples_per_frame;
	samples_this_frame = (UINT32)samples_left_over;
	samples_left_over -= (double)samples_this_frame;

	samples_this_frame += current_adjustment;

	// return the samples to play this next frame
	return samples_this_frame;
}

from the original mame source

**Comment by dankcushions** [comment]
yeah the libretro api has to handle the audio and visual syncing so the translation to that is likely where the issue is.

we had an issue in that resampler code with the MK driver, although it was fixed a few months ago, for that specific case.

**Comment by dankcushions** [comment]
which original setting?

**Comment by grant2258** [comment]
is there anywhere the api is documented whats the difference between timing.fps and 

  struct retro_system_timing t = {
      Machine->drv->frames_per_second,
      32000.0
   };

like mame2000 does i cant seem to find the api docs at all just a brief emulator example

**Comment by markwkidd** [comment]
The API docs are not very detailed yet.

`libretro.h` is supposed to be self-documented but that only goes so far:
https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h

The fastest way to get an answer to that specific question is probably to
ask in the #programming libretro Discord channel:
https://discordapp.com/channels/184109094070779904/419255960465309700

Would you like me to pose it there?

As far as model cores: skeletor is a very simple libretro core created by
one of the primary developers: https://github.com/fr500/skeletor

FreeIntv is a complete-ish Intellivision emulator that is native to
libretro and therefore much simpler:
https://github.com/libretro/FreeIntv/tree/master/src

**Comment by grant2258** [comment]
cheers mark should be enough to get me started apreciated

**Comment by grant2258** [comment]
Well made some progress got bally games working the mame200 osd_code. I still think the underun/overun code needs added the bally framerates are off but its working

**Comment by ZappaUtopia** [comment]
Would your fix also fix some of the vector games and a few others that have the same symptoms (just hang with a volume bar at the bottom or can't coin up) like Atari games Battlezone & Star Wars ?

**Comment by Wilstorm** [comment]
@ZappaUtopia - I think the Bally/MCR games are more like Kick, Tapper, Tron, Spy Hunter, Journey and Demolition Derby (one of the few games besides vector based games I have in AdvMAME along with Xenophobe) since it's the only emulator that runs them.

**Comment by ZappaUtopia** [comment]
Yeah I know, but some Atari games and a few others give the same problem when starting (can't coin up or hang with a volume bar at the bottom). I was just asking if the fix implemented will also fix these. (Journey works anyway).

**Comment by grant2258** [comment]
@arcadez ill need to update this seperatly for us as it will change out libretro.c im just going to run mk3 for 20 minutes to see if all is ok there shoudl be really but helps to make sure then ill pass the changes along

**Comment by Wilstorm** [comment]
@ZappaUtopia - I was wondering if the comment was based on symptoms. I am not sure if it's correct way to look at it but I always think of the driver they are working on and what games it affects. If it's a different driver then a separate issue might help them track it properly. Unless it was global or something like samples...nice catch! You probably already know but AdvMAME really shines when it comes to vector based games like Battlezone & Star Wars. Nothing really compares at this point but I think Mark has a keen eye for a fixer.

**Comment by ZappaUtopia** [comment]
@grant2258 I followed the thread that @windale reported over a year ago that you posted about here https://github.com/libretro/mame2003-libretro/issues/75 . It was about some games playing at 60 Hz when they weren't supposed to (Robocop was one of them which should be 57Hz) and it makes the game and sound play slightly too fast. Their workaround rather than solution was to set `audio_max_timing_skew = 0.02`. It seemed most people were in favour of playing them at the wrong speed but smooth 60Hz. I personally prefer the accurate speed like windale. Is this what it's all about or something completely different ?

**Comment by grant2258** [comment]
well the audio handling is what throwing the fps timer off with robocop. The code still had been changed to fix this 100%. The main thing is to get the inputs workings again we know where the syncing issues are now if you disable the audio cb robocop will runs at 57 fps. The mortal combat fix was just part of the mame code above i posted implemented. Unless there are other timing issues this will be fixed the mame code is available if someone beats me to it

**Comment by grant2258** [comment]
@arcadez I made the changes here try the dames your having input problems with. There is still a fps issue in some games but thats not an issue added its always been there.

Please only report games with sound that did work on mame 2003 and doesnt now. I donty think there will be any but never say never

**Comment by dankcushions** [comment]
@ZappaUtopia 
> @grant2258 I followed the thread that @windale reported over a year ago that you posted about here libretro/mame2003-libretro#75 . It was about some games playing at 60 Hz when they weren't supposed to (Robocop was one of them which should be 57Hz) and it makes the game and sound play slightly too fast. Their workaround rather than solution was to set audio_max_timing_skew = 0.02. It seemed most people were in favour of playing them at the wrong speed but smooth 60Hz. I personally prefer the accurate speed like windale. Is this what it's all about or something completely different ?

it's not a "workaround", it IS the solution that specific issue. you must understand that it is simply how libretro works (probably its best feature, even) - it handles audio/video sync. as part of that, it is able to skew the framerate to match up with your display refresh rate so you don't get judder (within the tolerance specified by audio_max_timing_skew) and resample the the audio in tandem. if you don't want it to do that, you simply set audio_max_timing_skew to something lower/0.

there is no fix to be done to the cores for that issue, nor should any libretro callbacks be avoided to bypass this, or similar functionality (not sure if that's what you've done, @grant2258 ?).

**Comment by grant2258** [comment]
i havent avoided any callbacks at all you can see the code ive changed

**Comment by dankcushions** [comment]
> @dankcushions Just out of interest, when you say there is no fix to be done to correct the fps, how come MAME 2010 plays them at the correct speed (Robocop etc.) ?. Just curious.

assuming the retroarch configuration is the same in both setups, maybe mame2010 is failing to update info->timing.fps when it loads a games with a fps that is not 60? in the code it appears this value defaults to 60: https://github.com/libretro/mame2010-libretro/blob/545acf2ef0031ef6f76acbaab4e37737b6fa8084/src/osd/retro/retromain.c#L764 and https://github.com/libretro/mame2010-libretro/blob/545acf2ef0031ef6f76acbaab4e37737b6fa8084/src/osd/retro/retromain.c#L68

it looks like it's attempting to update it here: https://github.com/libretro/mame2010-libretro/blob/545acf2ef0031ef6f76acbaab4e37737b6fa8084/src/osd/retro/retromain.c#L394 - but i'd have to debug it to find out if it's getting the expected values.

hmm, also it is using the RETRO_ENVIRONMENT_SET_GEOMETRY callback which may not be sufficient to update timings. i think it might need to call RETRO_ENVIRONMENT_SET_SYSTEM_AV_INFO instead. interesting!

if you lower the skew it will play the games at the correct framerate in mame2003.

**Comment by grant2258** [comment]
the timing changes when something is out of sync in the retro_run 

robocop should be running at 57fps go into the retroach menu in mame2003 leave it sit there for a while no samples will be playing and it will settle at 57fps

**Comment by dankcushions** [comment]
@grant2258 see https://github.com/libretro/mame2003-libretro/issues/75#issuecomment-272181047
and https://github.com/libretro/RetroArch/issues/4662

there was previously some weirdness where what it would show in the menu did not represent what the game was actually running at.

**Comment by grant2258** [comment]
its all down to how the samples are handled as far as i can see. I did mention earlier  I thought this was all related earlier . The reason the menu re syncs up is because there is no miss timings in there.

**Comment by ZappaUtopia** [comment]
So lr-mame2003 should be using lr-mame2010's method which is correct.

**Comment by grant2258** [comment]
it wont make a difference the retro_run audio loop is what throwing things off

**Comment by grant2258** [comment]
ok some details i found the problem with the sound on the midway games and its fixed easily by setting your sample rate to 22050.  The default 48000 is no good for anything with a framerate below 48. So if your having problems with games with fps below 48 you need to change the sample rate to 22050. Note for frame rate settings to take you need to restart retroarch. Now that problem is out the way to fix the fps problem display on robocop we need to run the core at 60fps I dont think any games go higher than that


any problems please feel free to revert and let me know how you get on. This could possibly fix the vector issues as well :)

**Comment by grant2258** [comment]
mame2010 does it like this 
static double refresh_rate = 60.0;
info->timing.fps            = refresh_rate;
info->timing.sample_rate = (double)sample_rate;

**Comment by dankcushions** [comment]
yeah like i said, different mame cores do it differently (libretro mame current does it the 2010 way also), and it might be the right way. maybe mame standalone is always dealing with an internal 60hz timing cycle, but then 'corrects' it later?

i'm leaning towards just going with the 2010 way and seeing what happens :)

**Comment by grant2258** [comment]
well we will just leave it your way dankcushions the lastest drops the samplerate   its too high . Ill leave the timing side to those who know it does work both ways though and it appears to run the same

**Comment by grant2258** [comment]
I dont think mame is tied to 60hz at all it changes its timing to keep things synced in the audio and  video vsunc ect. 

Mame is streaming the samples we a batch processing them that could be the big difference. We are only supposed to process the last frames samples   on osd_update and set the new one up for the next call. Looks like we are just batch processing it after its buffered

**Comment by ZappaUtopia** [comment]
Regarding the issue of some games playing at the wrong speed (i.e. Robocop) IMHO I think this should be corrected so that they play the speed it was intended. We shouldn't have to 'manually' change the 'audio timing skew' to correct this. This is the only MAME core that does it. It isn't even documented anywhere that you have to change the timing skew, only on Github.

**Comment by grant2258** [comment]
@ZappaUtopia  dont worry about what the fps says its playing right

**Comment by Wilstorm** [comment]
> Bally and Atari games are now working great (plus some others), I don't even notice the difference in sound quality.

Do you have a list of ROMs that have been corrected or the ones you have tested?

I always wondered about the sample rate with arcade games. The speakers we use for playback are so crappy that I wonder if it really makes a difference. The sample rate for practical purposes determines the max frequency. A 44.1kHz sample rate is around 20 kHz maximum frequency (human hearing max) but some folks like a sonar tech or young people can go higher. Hence those apps kids use nowadays that use audible frequencies outside of adults hearing.

Anyway at 22kHz you're talking a frequency range of 10KHz. Basically it's worse than a 128kbps mp3 but hence small tinny speakers don't really capture full range audio very well anyway.

**Comment by grant2258** [comment]
just because you set the sample rate to 48k its doesnt mean the hardware uses that sample rate

**Comment by grant2258** [comment]
I dont really want to get into the timing thing thats something you need to discuss with dan

**Comment by ZappaUtopia** [comment]
With the default timing skew Robocop plays at 60 Hz. It correctly reports it in the RetroArch menu as 57.4 but as soon as you start a game, it goes up to 60 Hz. You can hear the difference with the music, it's faster and the speech is a higher pitch. If you manually set the timing skew to 0.02 it corrects it but we shouldn't have to do this.

**Comment by Wilstorm** [comment]
I guess I was speaking audio in general. The setup I have on my PC packs a wallop for sure. I can rock the whole house. I am not sure why but Iron Maiden has been my go to these past few weeks. It changes week to week. I was really looking forward to some Rush 2112 in Ready Player One but it didn't happen.

Is there a general guide line on what it does use so we have some idea. Not that it matters but more for informational purposes. 

I think if Zappa would share just the ROM list name of the games tested that would be very helpful.

I trust your ability with programming completely. No need to discuss timing here. I am not sure what level of programmer Dank is but I suspect he can take a crack at fixing it since it wasn't quite the solution he was looking for. I was ok with the solution you proposed but talking mechanics of how something works and actually programming it are two different things entirely. It's the difference between 1's and 0's, night and day. peanut butter and jelly, Laurel and Hardy, ok so DUO's can go on forever... ;)

**Comment by grant2258** [comment]
I really dont know where Dan is coming from with this. I know why its doing it and why dropping the sample rate fixed it. 

It is fixed properly I dont know why he said it isint he will need to clarify that from his retroarch perspective to explain it to you.

**Comment by dankcushions** [comment]
@ZappaUtopia if you don’t like retroarch’s default timing skew you should go to them about it - there’s a topic on this very subject that mark linked to somewhere.

we shouldn’t bypass the api for stuff like this. please drop it.

**Comment by Wilstorm** [comment]
> I really dont know where Dan is coming from with this.

Oh his name is Dan. Ok so that just dawned on me. Hmmm...I am not sure but online you really never speak out loud and I always thought Dank.

**Comment by grant2258** [comment]
@markwkidd  the thing is mark its just the the way the audio is processed that causing the framerate to change its not a hack its to fit into the complementation thats been done for mame2003.  is the same as lr-mame but lrmame is cleaver they set a sensable sample rate and didnt make it definable

**Comment by Wilstorm** [comment]
> please drop it.

Ok, sorry, I'm not trying to ruffle feathers but I'm a huge advocate for freedom of speech in all forms and mediums even if it goes in a circle. Some new ideas will come from this I am sure. It takes two to have a conversation after all. 

If Zappa feels a need to share or ask questions and it looks like Grant was trying to understand or get some clarity too and he understands it on a another level completely. To attempt a change on a project he and two others mainly do all the programming on no less so why not? I thought that was why mame2003_plus_libretro came to fruition in the first place. For these types of scenarios.

I get the part of not wanting to participate that makes sense I guess though so...I guess that's good enough.

@ZappaUtopia - Ok, no big deal, if I am looking to play a game it will work or not but you do need to set the sample rate to 22050 kHz to get them working?

**Comment by grant2258** [comment]
you dont need to set anything on mame2003 plus

**Comment by Wilstorm** [comment]
Oh, nice, thanks Grant! Hopefully will do some testing this weekend.

**Comment by markwkidd** [comment]
I've also been reading this thread about RetroArch's audio skewing feature to try to educate myself on what's happening https://forums.libretro.com/t/perfect-audio-video-synchronization/12072

**Comment by grant2258** [comment]
mark lets look at it when it before it broke the pls is the changes that broke it.

 -  audio_batch_cb(XsoundBuffer, 800); 
+    audio_batch_cb(XsoundBuffer, Machine->sample_rate / Machine->drv->frames_per_second);

- info->timing.fps = videoConfig.fps;
+    info->timing.fps = Machine->drv->frames_per_second;


now lets look whats happens with tapper before and after

before the change
fps is 60 or around there 
sample rate is 48k (this timing works doable)
48000 / 60 = 800

now the change
fps = 30 (doable)
sample rate = 48000 / 30 = 1600 (not doable in this timeframe) change the timing

**Comment by grant2258** [comment]
Another thing i would like to add is I would like these cores kept compatible with each other. Thats why im not changing ours to my way. The cores should be able to fix each other not work against each other.

**Comment by grant2258** [comment]
I do mean this in technical terms @dankcushions and not on a personal level. The hint is in the functions  names osd operatng system dependent. I just think you got your thinking mixed up on what was going on internally with mame. One big issue was the cb wasnt called from osd_update as well. If you can live with the way it is I can as well the bottom line is it works.

**Comment by dankcushions** [comment]
I have said several times that it's not "my way" or anything like that. i'm just explaining why the api would normally want the correct fps being set via the av_info cbs. as we can see some other mame cores have just set this to 60, so perhaps that is the most sensible way. i've never said i wanted it a particular way - it's just the part i'm familiar with, and i can see an effect when we 'bypass' it.

you've lost me on the audio/sample rate stuff - i know next to nothing about audio programming but it sounds like the way we're interacting with the audio cbs has been wrong from the start and i would be surprised if there wasn't a 'correct' way of hooking in to the api, but i wouldn't know where to start with that, so do it the way you prefer!

**Comment by grant2258** [comment]
I know what you mean Dank I just can seem to find a way round the 30fps without dropping the framerate. Ill need to see how these games are running on mame2000 and mame2010 when they are 30fps

**Comment by grant2258** [comment]
this is how i would do it personally thoughts @dankcushions 

https://github.com/grant2258/mame2003-libretro/commit/e3c61dac3432a88dd974ff3813ee2169d48c610f

**Comment by grant2258** [comment]
going to test on linux windows and pi then ill do a pull request if your ok with it for here and mame2000
